File: Assignment 1/2020BTEIT00034/quickSort.py
"""
    Quick Sort
    ----------
    Daulatrao Patil - 2020BTEIT00034
    ----------

    Time Complexity: O(n log n)
    Space Complexity: O(log n)
    Stable: Yes

    Algorithm:
        1. Pick a pivot element
        2. Partition the array around the pivot element
        3. Recursively sort the subarrays

        Partition:
        --------------------
            1. Pick the last element as pivot
            2. Compare the pivot with all the elements in the array
            3. If the pivot is greater than the element, then move the pivot to the right
            4. If the pivot is less than the element, then move the pivot to the left
            5. Repeat the steps 2-4 until the pivot is at the right place
            6. Return the pivot position
"""

import random

def partition(A, low, high):
    pivot = A[high]
    i = low - 1

    for j in range(low, high):
        if A[j] <= pivot:
            i += 1
            A[i], A[j] = A[j], A[i]

    A[i+1], A[high] = A[high], A[i+1]

    return i+1

def quickSort(A, low, high):
    if low < high:
        pivot_ele = partition(A, low, high)

        quickSort(A, low, pivot_ele-1)
        quickSort(A, pivot_ele+1, high)
    else:
        return

def main():
    n = int(input())

    A = [random.randint(0, 1000) for i in range(n)]

    print("Unsorted array: ", A)

    quickSort(A, 0, n-1)

    print("Sorted array: ", A)
    
if __name__ == "__main__":
    main()

File: Assignment 1/2020BTEIT00051/quickSort.cpp
#include <iostream>
using namespace std;
  
void swap(int a, int b)      //function for swapping two elements.
{ 
    int t = a; 
    a = b; 
    b = t; 
} 
   
int partition (int arr[], int low, int high)     //function for partitioning the given array.
{ 
    int pivot = arr[high];   
    int i = (low - 1);   
   
    for (int j = low; j <= high- 1; j++) 
    { 
        if (arr[j] <= pivot) 
        { 
            i++;  
            swap(arr[i], arr[j]); 
        } 
    } 
    swap(arr[i + 1], arr[high]); 
    return (i + 1); 
} 
   
void quickSort(int arr[], int low, int high)      //driver function for quicksort.
{ 
    if (low < high) 
    { 
        int pivot = partition(arr, low, high); 
   
        quickSort(arr, low, pivot - 1); 
        quickSort(arr, pivot + 1, high); 
    } 
} 
   
void displayArray(int arr[], int size) 
{ 
     
    for (int i=0; i < size; i++) 
        cout<<arr[i]<<"\t"; 
      
} 
   
int main() 
{ 
    int arr[1000];
    for(int i=0;i<1000;i++){
        arr[i]=10000-i;
    }
    int n = sizeof(arr)/sizeof(arr[0]); 
    cout<<"Input array is :"<<endl;
    displayArray(arr,n);
    cout<<endl;
    quickSort(arr, 0, n-1); 
    cout<<"Array sorted with quick sort :"<<endl; 
    displayArray(arr,n); 
    return 0; 
}
